### ISSUE TITLE
Destructors

### ISSUE THREAD

**Issue by defval** [defval/inject#22]
Add new provide option `OnDestruct()`.

```go
// OnDestruct is a provide option, which modify default providing behavior.
// It adds destructor callback function.
func OnDestruct(fn interface) ProvideOption
```

Add new container function `Destruct(),` which destructs all initialized instances in the container.

```go
// Destruct is a container function, that destruct all initialized instances.
func (c *Container) Destruct() error
```

Example usage:

```go
// StopServer is a callback function that will be called with a container destruction.
func StopServer(server *http.Server) error
```

```go
container, err := inject.New(
  inject.Provide(NewHTTPServer, inject.OnDestruct(StopServer))
)

if err != nil {
  panic(err)
}

// do stuff

if err = container.Destruct(); err != nil {
  panic(err)
}
```

### TBD:

- Timeouts

**Comment by maxzhang1985** [comment]
I think used interface better. such as  IDestruct interface has a OnDestruct method.

**Comment by defval** [comment]
@maxzhang1985 I'd prefer to separate the object lifecycle management functions from the object functionality. Also, interface forces to wrapping standard library types. 

I start to implement the [Provider](https://github.com/defval/inject/blob/v2/di/provider.go#L6) interface in `v2` for advanced providing. Maybe, extended `ProviderDestructor` or `ProviderCloser` interface will be the best option.

```go
type Provider interface {
    Identity() Identity
    Construct() (interface, error)
}

type ProviderDestructor interface {
    Destruct() error
}
```

I didn't have time to test it. But here a usage [example](https://github.com/defval/inject/blob/v2/di/container_test.go#L211).

I'm still thinking about it.

**Comment by defval** [comment]
I came to the conclusion that a better option is to support an optional cleanup function as a result of the constructor (like in [google/wire](https://github.com/google/wire/blob/master/docs/guide.md#cleanup-functions)).

For example: 

```go
func provideFile(log Logger, path Path) (*os.File, func(), error) {
    f, err := os.Open(string(path))
    if err != nil {
        return nil, nil, err
    }
    cleanup := func() {
        if err := f.Close(); err != nil {
            log.Log(err)
        }
    }
    return f, cleanup, nil
}
```

**Comment by defval** [comment]
Implemented in #26
